spec/core/ics-026-routing-module/UPGRADES.md

```diff
@@ -0,0 +1,141 @@
+are valid and perform any custom `UpgradeInit` logic.
+It may return an error if the chosen parameters are invalid 
+in which case the upgrade handshake is aborted.
+If the provided version string is empty, `onChanUpgradeInit` should return 
+a default version string or an error if the provided version is invalid.
+Note that if the upgrade provides an empty string, this is an indication to upgrade
+to the default version which MAY be a new default from when the channel was first initiated.
+If there is no default version string for the application,
+it should return an error if provided version is empty string.
+
+`onChanUpgradeInit` is also responsible for making sure that the application is recoverable to its pre-upgrade state. The application may either store any new metadata in separate paths, or store the previous metadata under a different path so it can be restored.
+
+```typescript
+function onChanUpgradeInit(
+  order: ChannelOrder,
+  connectionHops: [Identifier],
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier,
+  counterpartyPortIdentifier: Identifier,
+  counterpartyChannelIdentifier: Identifier,
+  version: string) => (version: string, err: Error) {
+    // defined by the module
+}
+```
+
+#### **OnChanUpgradeTry**
+
+`onChanUpgradeTry` will verify the upgrade-chosen parameters and perform custom `TRY` logic. 
+If the upgrade-chosen parameters are invalid, the callback must return an error to abort the handshake. 
+If the counterparty-chosen version is not compatible with this modules
+supported versions, the callback must return an error to abort the handshake. 
+If the versions are compatible, the try callback must select the final version
+string and return it to core IBC.
+`onChanUpgradeTry` may also perform custom initialization logic.
+
+`onChanUpgradeTry` is also responsible for making sure that the application is recoverable to its pre-upgrade state. The application may either store any new metadata in separate paths, or store the previous metadata under a different path so it can be restored.
+
+```typescript
+function onChanUpgradeTry(
+  order: ChannelOrder,
+  connectionHops: [Identifier],
+  portIdentifier: Identifier,
+  channelIdentifier: Identifier,
+  counterpartyPortIdentifier: Identifier,
+  counterpartyChannelIdentifier: Identifier,
+  counterpartyVersion: string) => (version: string, err: Error) {
+    // defined by the module
+}
+```
+
+#### **OnChanUpgradeAck**
+
+`onChanUpgradeAck` will error if the counterparty selected version string
+is invalid to abort the handshake. It may also perform custom ACK logic.
+
+After `onChanUpgradeAck` returns, the application upgrade is complete on this end so any 
+auxilliary data stored for the purposes of recovery is no longer needed and may be deleted.
+
+The application MUST have its state fully migrated to start processing packet data according to the new application parameters by the time the callback returns.
```

Comment:
So the application parameters are any parameters that application needs to run the application logic that is needed for the channel.

This will be the channel parameters like version, but it may also be completely internal parameters defined and maintained by the end application.

These parameters might be created by the application callbacks during the channel upgrade process